mongodb/src/main/java/org/hibernate/ogm/datastore/mongodb/dialect/impl/EmbeddableStateFinder.java

```diff
@@ -0,0 +1,132 @@
+/*
+ * Hibernate OGM, Domain model persistence for NoSQL datastores
+ *
+ * License: GNU Lesser General Public License (LGPL), version 2.1 or later
+ * See the lgpl.txt file in the root directory or <http://www.gnu.org/licenses/lgpl-2.1.html>.
+ */
+
+package org.hibernate.ogm.datastore.mongodb.dialect.impl;
+
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+import org.hibernate.ogm.dialect.spi.TupleContext;
+import org.hibernate.ogm.model.spi.Tuple;
+
+/**
+ * Captures useful data around the state of embeddable objects in tuples
+ * Note that the current implementation is *not* specific to MongoDB on purpose.
```

Comment:
ah yes I did not think of this approach, we can do that. Keeping it in impl but shared in core.